### ISSUE TITLE
Segmentation fault with GPU expressions

### ISSUE THREAD

**Issue by stefanwebb** [pytorch/pytorch#6623]
Hi,
I am implementing a probabilistic model that makes use of the Gamma distribution. Doing a `.rsample()` on the GPU however generate a segmentation fault, whereas doing it on the CPU is fine.

Is this a bug or is GPU Gamma sampling unimplemented?

Thanks!

**Comment by apaszke** [comment]
Can you please send us a small example code that would let us reproduce the issue?

**Comment by stefanwebb** [comment]
Thanks for the quick reply! The following gives me `Segmentation fault (core dumped)`

>import numpy as np
import torch
from torch.autograd import Variable
import torch.distributions as dist
t_a = Variable(torch.Tensor([0.1, 0.3, 0.5])).cuda()
t_b = Variable(torch.Tensor([0.6, 0.7, 0.8])).cuda()
t_dist = dist.Gamma(t_a, t_b)
t_sample = t_dist.rsample()
print(t_sample.cpu().numpy())`

I'm running PyTorch commit `084e3a7` from 12 hours ago

**Comment by stefanwebb** [comment]
I think this is a more serious problem. I am trying to implement my own PyTorch distribution for the exponential concrete distribution but am running into the same problem. When I add together two expressions in the `.log_prob` function that are on the GPU it gives a segmentation fault:

> log_norm_const = lgamma(k) + (k - 1.) * torch.log(self.temperature)
log_softmax = torch.nn.LogSoftmax(dim=-1)(self.logits - value * self.temperature)
log_prob = log_norm_const + log_softmax

But if I combine them into a single statement it is okay:

> log_softmax = torch.nn.LogSoftmax(dim=-1)(self.logits - value * self.temperature) + lgamma(k) + (k - 1.) * torch.log(self.temperature)

**Comment by apaszke** [comment]
Hmm that's weird. I just tried the repro you posted and it gave me
```
Traceback (most recent call last):
  File "tmp2.py", line 8, in <module>
    t_sample = t_dist.rsample()
  File "/home/apaszke/pytorch/torch/distributions/gamma.py", line 55, in rsample
    value = _standard_gamma(self.concentration.expand(shape)) / self.rate.expand(shape)
  File "/home/apaszke/pytorch/torch/distributions/gamma.py", line 12, in _standard_gamma
    return concentration._standard_gamma()
RuntimeError: _standard_gamma is not implemented for type torch.cuda.FloatTensor
```
which is reasonable (and will be fixed by #6855). Can you please try that again with a newer commit?

**Comment by stefanwebb** [comment]
It all seems to work now (including on the GPU) with the latest commit
